src/cluster.py

```diff
@@ -0,0 +1,212 @@
+        if not event.unit:
+            return
+        cmd = ["sudo", "microceph", "cluster", "add", event.unit.name]
+        try:
+            logger.debug(f'Running command {" ".join(cmd)}')
+            process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
+            logger.debug(f"Command finished. stdout={process.stdout}, " f"stderr={process.stderr}")
+            token = process.stdout.strip()
+            self.charm.peers.set_app_data({f"{event.unit.name}.join_token": token})
+        except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
+            logger.warning(e.stderr)
+            error_node_already_exists = (
+                'Failed to create "internal_token_records" entry: UNIQUE '
+                "constraint failed: internal_token_records.name"
+            )
+            if error_node_already_exists not in e.stderr:
+                raise e
+
+    def join_node_to_cluster(self, event: ops.framework.EventBase) -> None:
+        """Join node to microceph cluster."""
+        if not event.unit:
+            return
+        token = self.charm.peers.get_app_data(f"{event.unit.name}.join_token")
+        cmd = ["sudo", "microceph", "cluster", "join", token]
+        try:
+            logger.debug(f'Running command {" ".join(cmd)}')
+            process = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=180)
+            logger.debug(f"Command finished. stdout={process.stdout}, " f"stderr={process.stderr}")
+            self.charm.peers.interface.state.joined = True
+            self.charm.peers.set_unit_data({"joined": json.dumps(True)})
+        except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
+            logger.warning(e.stderr)
+            raise e
+
+
+class ClusterUpgrades(ops.framework.Object):
+    """ClusterUpgrades manages snap upgrades across the cluster."""
+
+    charm = None
+    _stored = ops.framework.StoredState()
+
+    def __init__(self, charm: "charm.MicroCephCharm"):
+        super().__init__(charm, "cluster-upgrade")
+        self.charm = charm
+
+    @property
+    def peer_int(self):
+        """Get the peer interface."""
+        return self.charm.peers.interface
+
+    @property
+    def channel(self) -> str:
+        """Get the snap channel."""
+        return self.charm.channel
+
+    @channel.setter
+    def channel(self, value: str) -> None:
+        self.charm.channel = value
+
+    @property
+    def track(self) -> str:
+        """Get the snap track."""
+        c = self.channel
+        if not self.channel:
+            c = self.model.config["snap-channel"]
+        return c.split("/")[0]
+
+    def upgrade_requested(self, chan: str) -> bool:
+        """Check if a snap upgrade was requested."""
+        logger.debug(f"Requested, current channel: {chan}, {self.channel}")
+        return self.channel != chan
+
+    def can_upgrade_snap(self, snap_chan: str) -> Tuple[bool, str]:
```

Comment:
I see that `microceph.py` also has a method with same name (i.e. `can_upgrade_snap`) and I wonder if (for clarity) it would make sense to change this method's name to something that implies the charm context (e.g. `can_upgrade_charm_payload` etc.)